aws/resource_aws_codepipeline.go

```diff
@@ -109,35 +104,24 @@ func resourceAwsCodePipeline() *schema.Resource {
 							Elem: &schema.Resource{
 								Schema: map[string]*schema.Schema{
 									"configuration": {
-										Type:             schema.TypeMap,
-										Optional:         true,
-										Elem:             &schema.Schema{Type: schema.TypeString},
-										DiffSuppressFunc: suppressCodePipelineStageActionConfiguration,
```

Comment:
To prevent plan differences on provider upgrade, the old difference suppression needs to be in place, otherwise practitioners will unexpectedly receive this plan output with the sensitive value in plaintext:

```
  # aws_codepipeline.test will be updated in-place
  ~ resource "aws_codepipeline" "test" {
        id       = "bflad-testing"
        name     = "bflad-testing"
        tags     = {}
        # (2 unchanged attributes hidden)


      ~ stage {
            name = "Source"

          ~ action {
              ~ configuration    = {
                  ~ "OAuthToken" = "hash-OMITTED" -> "OMITTED"
                    # (3 unchanged elements hidden)
                }
                name             = "Source"
                # (7 unchanged attributes hidden)
            }
        }
        # (2 unchanged blocks hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.

Warning: The CodePipeline GitHub version 1 action provider is deprecated.

  on main.tf line 36, in resource "aws_codepipeline" "test":
  36:       provider         = "GitHub"

Use a GitHub version 2 action (with a CodeStar Connection
`aws_codestarconnections_connection`) instead. See
https://docs.aws.amazon.com/codepipeline/latest/userguide/update-github-action-connections.html
```

This suppression would not prevent updates from receiving the correct (non-hashed) value and applying it to the state as expected. We did this recently with `aws_acm_certificate` resource, for an example.

Tested via applying with version 3.23.0 and then using [development overrides](https://www.terraform.io/docs/commands/cli-config.html#development-overrides-for-provider-developers).

```terraform
terraform {
  required_version = "0.14.4"
  required_providers {
    aws = {
      source  = "hashicorp/aws"
      version = "3.23.0"
    }
  }
}

provider "aws" {
  region = "us-east-2"
}

resource "aws_codepipeline" "test" {
  name     = "bflad-testing"
  role_arn = aws_iam_role.test.arn

  artifact_store {
    location = aws_s3_bucket.test.bucket
    type     = "S3"

    encryption_key {
      id   = "1234"
      type = "KMS"
    }
  }

  stage {
    name = "Source"

    action {
      name             = "Source"
      category         = "Source"
      owner            = "ThirdParty"
      provider         = "GitHub"
      version          = "1"
      output_artifacts = ["test"]

      configuration = {
        Owner      = "lifesum-terraform"
        Repo       = "test"
        Branch     = "main"
        OAuthToken = "OMITTED :)"
      }
    }
  }

  stage {
    name = "Build"

    action {
      name            = "Build"
      category        = "Build"
      owner           = "AWS"
      provider        = "CodeBuild"
      input_artifacts = ["test"]
      version         = "1"

      configuration = {
        ProjectName = "test"
      }
    }
  }
}

resource "aws_s3_bucket" "test" {
  bucket = "bflad-testing"
  acl    = "private"
}

resource "aws_iam_role" "test" {
  name = "bflad-testing"

  assume_role_policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "codepipeline.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF
}

resource "aws_iam_role_policy" "test" {
  name = "codepipeline_policy"
  role = aws_iam_role.test.id

  policy = <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "s3:GetObject",
        "s3:GetObjectVersion",
        "s3:GetBucketVersioning"
      ],
      "Resource": [
        "${aws_s3_bucket.test.arn}",
        "${aws_s3_bucket.test.arn}/*"
      ]
    },
    {
      "Effect": "Allow",
      "Action": [
        "codebuild:BatchGetBuilds",
        "codebuild:StartBuild"
      ],
      "Resource": "*"
    }
  ]
}
EOF
}
```

Reply:
Verified this works as expected now! 👍